File: taikyoku_engine/movegen.py
"""Move generation for Taikyoku Shogi."""

from .pieces import (
    BOARD_SIZE, BLACK, WHITE, MOVEMENTS, PROMOTES_TO,
    ROYAL_PIECES, PIECE_RANK, RANK_ROYAL, RANK_GREAT, RANK_VICE,
    get_deltas, DELTAS, N, NE, E, SE, S, SW, W, NW, ALL_DIRS,
)
from .move import Move
from .board import TaikyokuBoard


def generate_pseudo_legal_moves(board: TaikyokuBoard):
    """Generate all pseudo-legal moves for the side to move.

    Does NOT filter out moves that leave own royals in check
    (Taikyoku Shogi doesn't have check rules - you can move into check,
     but losing all royals loses the game).
    """
    color = board.side_to_move
    moves = []

    for (r, c), piece in list(board.piece_positions[color].items()):
        movement = MOVEMENTS.get(piece)
        if movement is None:
            continue

        # Standard slides (step and range moves)
        _gen_slides(board, r, c, piece, color, movement, moves)

        # Jump moves
        _gen_jumps(board, r, c, piece, color, movement, moves)

        # Hook moves
        if movement.get('hook'):
            _gen_hooks(board, r, c, piece, color, movement, moves)

        # Area moves (lion-like)
        if movement.get('area', 0) > 0:
            _gen_area(board, r, c, piece, color, movement, moves)

        # Range capture
        if movement.get('range_capture'):
            _gen_range_capture(board, r, c, piece, color, movement, moves)

        # Igui
        if movement.get('igui'):
            _gen_igui(board, r, c, piece, color, moves)

    return moves


def generate_legal_moves(board: TaikyokuBoard):
    """Generate all legal moves. In Taikyoku Shogi, all pseudo-legal moves
    are legal since there's no check/checkmate rule.

    However, we should still filter out moves that are clearly suicidal
    (moving your last royal into capture) for better play.
    For now, return all pseudo-legal moves.
    """
    return generate_pseudo_legal_moves(board)


def _can_promote(piece):
    """Check if a piece can promote (has a promotion target)."""
    return piece in PROMOTES_TO


def _gen_slides(board, r, c, piece, color, movement, moves):
    """Generate slide/step moves along straight lines."""
    for direction, max_range in movement.get('slides', []):
        dr, dc = get_deltas(direction, color)
        if max_range == 0:
            max_range = BOARD_SIZE  # unlimited
        for dist in range(1, max_range + 1):
            nr, nc = r + dr * dist, c + dc * dist
            if not board.is_valid(nr, nc):
                break
            target = board.at(nr, nc)
            if target is None:
                # Empty square - can move here
                _add_move(moves, r, c, nr, nc, piece, color, None)
            elif target[1] != color:
                # Enemy piece - can capture
                _add_move(moves, r, c, nr, nc, piece, color, target)
                break  # Can't go further after capture
            else:
                # Friendly piece - blocked
                break


def _gen_jumps(board, r, c, piece, color, movement, moves):
    """Generate jump moves (bypassing intervening pieces)."""
    for jdr, jdc in movement.get('jumps', []):
        # Adjust for color
        if color == WHITE:
            jdr, jdc = -jdr, -jdc
        nr, nc = r + jdr, c + jdc
        if not board.is_valid(nr, nc):
            continue
        target = board.at(nr, nc)
        if target is None:
            _add_move(moves, r, c, nr, nc, piece, color, None)
        elif target[1] != color:
            _add_move(moves, r, c, nr, nc, piece, color, target)


def _gen_hooks(board, r, c, piece, color, movement, moves):
    """Generate hook moves (move along a line, then turn 90 degrees)."""
    hook_type = movement['hook']
    if hook_type == 'orth':
        # Hook on orthogonal lines: move along R, can turn 90 to continue on R
        dirs = [N, E, S, W]
    else:  # 'diag'
        dirs = [NE, SE, SW, NW]

    for d in dirs:
        dr, dc = get_deltas(d, color)
        # Move along primary direction
        for dist1 in range(1, BOARD_SIZE):
            nr, nc = r + dr * dist1, c + dc * dist1
            if not board.is_valid(nr, nc):
                break
            target = board.at(nr, nc)
            if target is not None:
                if target[1] != color:
                    # Can capture here (stop)
                    _add_move(moves, r, c, nr, nc, piece, color, target)
                break  # Blocked

            # Can stop here (already added by slides if applicable)
            # Now try turning 90 degrees at this point
            if hook_type == 'orth':
                turn_dirs = [E, W] if d in (N, S) else [N, S]
            else:
                turn_dirs_map = {
                    NE: [NW, SE], SE: [NE, SW],
                    SW: [SE, NW], NW: [NE, SW],
                }
                turn_dirs = turn_dirs_map.get(d, [])

            for td in turn_dirs:
                tdr, tdc = get_deltas(td, color)
                for dist2 in range(1, BOARD_SIZE):
                    nr2, nc2 = nr + tdr * dist2, nc + tdc * dist2
                    if not board.is_valid(nr2, nc2):
                        break
                    target2 = board.at(nr2, nc2)
                    if target2 is None:
                        _add_move(moves, r, c, nr2, nc2, piece, color, None)
                    elif target2[1] != color:
                        _add_move(moves, r, c, nr2, nc2, piece, color, target2)
                        break
                    else:
                        break


def _gen_area(board, r, c, piece, color, movement, moves):
    """Generate area moves (lion-like: multiple steps per turn)."""
    area_range = movement['area']
    # For area=2: can make up to 2 steps in any direction
    # This means it can reach any square within Manhattan distance 2
    # It can also capture one piece, then continue to an adjacent square

    # Simple implementation: generate all destination squares within range
    # For area=2: all squares reachable in 1 or 2 king steps
    visited = set()
    visited.add((r, c))

    # First step destinations
    for d in ALL_DIRS:
        dr, dc = get_deltas(d, color)
        r1, c1 = r + dr, c + dc
        if not board.is_valid(r1, c1):
            continue
        target1 = board.at(r1, c1)
        if target1 is not None and target1[1] == color:
            continue  # Blocked by friendly piece

        # Can reach (r1, c1) after first step
        if (r1, c1) not in visited:
            visited.add((r1, c1))
            if target1 is None:
                # Empty - can stop here or continue
                _add_move(moves, r, c, r1, c1, piece, color, None)
            else:
                # Enemy - capture and stop, or capture and continue
                _add_move(moves, r, c, r1, c1, piece, color, target1)

            # Second step (if area >= 2)
            if area_range >= 2:
                for d2 in ALL_DIRS:
                    dr2, dc2 = get_deltas(d2, color)
                    r2, c2 = r1 + dr2, c1 + dc2
                    if not board.is_valid(r2, c2):
                        continue
                    if (r2, c2) == (r, c):
                        # Returning to start - this is igui if captured on way
                        if target1 is not None and target1[1] != color:
                            m = Move(
                                from_sq=(r, c), to_sq=(r, c),
                                captured=target1[0], captured_color=target1[1],
                                mid_sq=(r1, c1), mid_captured=target1[0],
                                mid_captured_color=target1[1], is_igui=True
                            )
                            moves.append(m)
                        continue
                    target2 = board.at(r2, c2)
                    if target2 is not None and target2[1] == color:
                        continue
                    # Can reach (r2, c2) via (r1, c1)
                    if target1 is not None and target1[1] != color:
                        # Captured at mid-point
                        m = Move(
                            from_sq=(r, c), to_sq=(r2, c2),
                            mid_sq=(r1, c1), mid_captured=target1[0],
                            mid_captured_color=target1[1],
                        )
                        if target2 is not None:
                            m.captured = target2[0]
                            m.captured_color = target2[1]
                        moves.append(m)
                    elif target2 is not None and target2[1] != color:
                        # No capture at midpoint, capture at destination
                        _add_move(moves, r, c, r2, c2, piece, color, target2)
                    else:
                        # No captures along the way
                        _add_move(moves, r, c, r2, c2, piece, color, None)


def _gen_range_capture(board, r, c, piece, color, movement, moves):
    """Generate range capture moves (fly over pieces, capturing all).

    Range capturers fly over pieces of LOWER rank (higher numeric rank value),
    capturing all of them (friend or foe). They cannot fly over pieces of
    equal or higher rank (lower numeric value). They can also land on empty
    squares or on lower-rank pieces.
    """
    piece_rank = PIECE_RANK.get(piece, 4)

    for d in movement.get('range_capture', []):
        dr, dc = get_deltas(d, color)
        captured_list = []

        for dist in range(1, BOARD_SIZE):
            nr, nc = r + dr * dist, c + dc * dist
            if not board.is_valid(nr, nc):
                break

            target = board.at(nr, nc)
            if target is None:
                # Empty square - can land here
                m = Move(from_sq=(r, c), to_sq=(nr, nc),
                         range_captures=list(captured_list) if captured_list else None)
                moves.append(m)
            else:
                t_piece, t_color = target
                t_rank = PIECE_RANK.get(t_piece, 4)
                if t_rank > piece_rank:
                    # Lower rank piece - capture it and continue
                    captured_list.append((nr, nc, t_piece, t_color))
                    # Can also choose to land here
                    m = Move(from_sq=(r, c), to_sq=(nr, nc),
                             range_captures=list(captured_list))
                    moves.append(m)
                else:
                    # Equal or higher rank - blocked, cannot pass
                    break


def _gen_igui(board, r, c, piece, color, moves):
    """Generate igui moves (capture adjacent piece without moving)."""
    for d in ALL_DIRS:
        dr, dc = get_deltas(d, color)
        nr, nc = r + dr, c + dc
        if not board.is_valid(nr, nc):
            continue
        target = board.at(nr, nc)
        if target is not None and target[1] != color:
            m = Move(
                from_sq=(r, c), to_sq=(r, c),
                captured=target[0], captured_color=target[1],
                is_igui=True,
            )
            # Check for promotion on igui capture
            if _can_promote(piece):
                m2 = Move(
                    from_sq=(r, c), to_sq=(r, c),
                    captured=target[0], captured_color=target[1],
                    is_igui=True, promotion=True,
                )
                moves.append(m2)
            moves.append(m)


def _add_move(moves, fr, fc, tr, tc, piece, color, target):
    """Add a move, including promotion variant if applicable."""
    captured = target[0] if target else None
    cap_color = target[1] if target else None

    # Normal move
    m = Move(
        from_sq=(fr, fc), to_sq=(tr, tc),
        captured=captured, captured_color=cap_color,
    )
    moves.append(m)

    # Promotion variant: only when capturing (promotion by capture rule)
    if captured is not None and _can_promote(piece):
        m_promo = Move(
            from_sq=(fr, fc), to_sq=(tr, tc),
            promotion=True,
            captured=captured, captured_color=cap_color,
        )
        moves.append(m_promo)


def choose_random_move(board: TaikyokuBoard):
    """Choose a uniformly random legal move."""
    import random
    moves = generate_legal_moves(board)
    if not moves:
        return None
    return random.choice(moves)

File: src/movegen.rs
use crate::types::*;
use crate::pieces;
use crate::board::Board;

pub fn generate_legal_moves(board: &Board) -> Vec<Move> {
    let color = board.side_to_move;
    let c = color as usize;
    let rt = ray_table();
    let mut moves = Vec::with_capacity(512);

    for i in 0..board.piece_list_len[c] {
        let sq = board.piece_list[c][i] as usize;
        if sq == INVALID_SQ as usize { continue; }
        let cell = board.cells[sq];
        if cell == EMPTY_CELL { continue; }
        let pt = cell_piece(cell);
        let mv = pieces::movement(pt);

        gen_slides(board, sq, pt, color, mv, rt, &mut moves);
        gen_jumps(board, sq, pt, color, mv, &mut moves);

        if mv.hook.is_some() {
            gen_hooks(board, sq, pt, color, mv, rt, &mut moves);
        }
        if mv.area > 0 {
            gen_area(board, sq, pt, color, mv, &mut moves);
        }
        if !mv.range_capture.is_empty() {
            gen_range_capture(board, sq, pt, color, mv, rt, &mut moves);
        }
        if mv.igui {
            gen_igui(board, sq, pt, color, &mut moves);
        }
    }
    moves
}

#[inline]
fn can_promote(pt: u16) -> bool {
    pieces::promotes_to(pt).is_some()
}

/// Zone-based promotion rules:
/// 1. Entering the promotion zone from outside -> may promote
/// 2. Capturing inside the promotion zone -> may promote
/// 3. Reaching the farthest rank with a forward-only piece -> MUST promote
/// 4. Promotion is optional otherwise
fn add_move(moves: &mut Vec<Move>, from: u16, to: u16, pt: u16, color: u8, target: Cell) {
    let captured = if target != EMPTY_CELL { cell_piece(target) } else { 0 };
    let cap_color = if target != EMPTY_CELL { cell_color(target) } else { 0 };

    if !can_promote(pt) {
        // No promotion possible — just add the plain move
        moves.push(Move {
            from_sq: from, to_sq: to, promotion: false,
            captured_piece: captured, captured_color: cap_color,
            is_igui: false, mid_sq: INVALID_SQ, mid_piece: 0, mid_color: 0,
            range_caps: None,
        });
        return;
    }

    let from_in_zone = in_promo_zone(from as usize, color);
    let to_in_zone = in_promo_zone(to as usize, color);
    let is_capture = captured != 0;

    // Determine if promotion is allowed this move
    let may_promote =
        (!from_in_zone && to_in_zone) ||       // entering zone
        (from_in_zone && to_in_zone && is_capture); // capture inside zone

    // Determine if promotion is forced (forward-only piece at farthest rank)
    let must_promote = to_in_zone
        && is_farthest_rank(to as usize, color)
        && pieces::must_promote_at_far_rank(pt);

    if must_promote {
        // Only the promoted move
        moves.push(Move {
            from_sq: from, to_sq: to, promotion: true,
            captured_piece: captured, captured_color: cap_color,
            is_igui: false, mid_sq: INVALID_SQ, mid_piece: 0, mid_color: 0,
            range_caps: None,
        });
    } else if may_promote {
        // Both options: promote or don't
        moves.push(Move {
            from_sq: from, to_sq: to, promotion: false,
            captured_piece: captured, captured_color: cap_color,
            is_igui: false, mid_sq: INVALID_SQ, mid_piece: 0, mid_color: 0,
            range_caps: None,
        });
        moves.push(Move {
            from_sq: from, to_sq: to, promotion: true,
            captured_piece: captured, captured_color: cap_color,
            is_igui: false, mid_sq: INVALID_SQ, mid_piece: 0, mid_color: 0,
            range_caps: None,
        });
    } else {
        // No promotion available
        moves.push(Move {
            from_sq: from, to_sq: to, promotion: false,
            captured_piece: captured, captured_color: cap_color,
            is_igui: false, mid_sq: INVALID_SQ, mid_piece: 0, mid_color: 0,
            range_caps: None,
        });
    }
}

fn gen_slides(board: &Board, sq: usize, pt: u16, color: u8, mv: &Movement,
              rt: &RayTable, moves: &mut Vec<Move>) {
    for &(dir, max_range) in &mv.slides {
        let ray = rt.ray_for_color(sq, dir as usize, color);
        let limit = if max_range == 0 { ray.len() } else { (max_range as usize).min(ray.len()) };

        for j in 0..limit {
            let target_sq = ray[j] as usize;
            let target = board.cells[target_sq];
            if target == EMPTY_CELL {
                add_move(moves, sq as u16, target_sq as u16, pt, color, EMPTY_CELL);
            } else if cell_color(target) != color {
                add_move(moves, sq as u16, target_sq as u16, pt, color, target);
                break;
            } else {
                break; // friendly piece blocks
            }
        }
    }
}

fn gen_jumps(board: &Board, sq: usize, pt: u16, color: u8, mv: &Movement,
             moves: &mut Vec<Move>) {
    let r = sq_row(sq) as i32;
    let c = sq_col(sq) as i32;
    for &(jdr, jdc) in &mv.jumps {
        let (dr, dc) = if color == BLACK {
            (jdr as i32, jdc as i32)
        } else {
            (-(jdr as i32), -(jdc as i32))
        };
        let nr = r + dr;
        let nc = c + dc;
        if nr < 0 || nr >= BOARD_SIZE as i32 || nc < 0 || nc >= BOARD_SIZE as i32 { continue; }
        let nsq = nr as usize * BOARD_SIZE + nc as usize;
        let target = board.cells[nsq];
        if target == EMPTY_CELL {
            add_move(moves, sq as u16, nsq as u16, pt, color, EMPTY_CELL);
        } else if cell_color(target) != color {
            add_move(moves, sq as u16, nsq as u16, pt, color, target);
        }
    }
}

fn gen_hooks(board: &Board, sq: usize, pt: u16, color: u8, mv: &Movement,
             rt: &RayTable, moves: &mut Vec<Move>) {
    let dirs: &[usize] = match mv.hook {
        Some(HookType::Orthogonal) => &[N, E, S, W],
        Some(HookType::Diagonal) => &[NE, SE, SW, NW],
        None => return,
    };

    for &d in dirs {
        let ray = rt.ray_for_color(sq, d, color);
        for (_j, &mid_sq) in ray.iter().enumerate() {
            let mid = mid_sq as usize;
            let target = board.cells[mid];
            if target != EMPTY_CELL {
                if cell_color(target) != color {
                    add_move(moves, sq as u16, mid_sq, pt, color, target);
                }
                break;
            }
            // At this empty square, try turning 90 degrees
            let turn_dirs = match mv.hook {
                Some(HookType::Orthogonal) => {
                    if d == N || d == S { vec![E, W] } else { vec![N, S] }
                }
                Some(HookType::Diagonal) => {
                    match d {
                        NE => vec![NW, SE], SE => vec![NE, SW],
                        SW => vec![SE, NW], NW => vec![NE, SW],
                        _ => vec![],
                    }
                }
                None => vec![],
            };
            for td in turn_dirs {
                let turn_ray = rt.ray_for_color(mid, td, color);
                for &tsq in turn_ray {
                    let t = board.cells[tsq as usize];
                    if t == EMPTY_CELL {
                        add_move(moves, sq as u16, tsq, pt, color, EMPTY_CELL);
                    } else if cell_color(t) != color {
                        add_move(moves, sq as u16, tsq, pt, color, t);
                        break;
                    } else {
                        break;
                    }
                }
            }
        }
    }
}

fn gen_area(board: &Board, sq: usize, pt: u16, color: u8, mv: &Movement,
            moves: &mut Vec<Move>) {
    let r = sq_row(sq) as i32;
    let c = sq_col(sq) as i32;

    for d1 in 0..NUM_DIRS {
        let (dr1, dc1) = get_deltas(d1, color);
        let r1 = r + dr1;
        let c1 = c + dc1;
        if r1 < 0 || r1 >= BOARD_SIZE as i32 || c1 < 0 || c1 >= BOARD_SIZE as i32 { continue; }
        let sq1 = r1 as usize * BOARD_SIZE + c1 as usize;
        let t1 = board.cells[sq1];
        if t1 != EMPTY_CELL && cell_color(t1) == color { continue; }

        // First step destination
        add_move(moves, sq as u16, sq1 as u16, pt, color, t1);

        // Second step
        if mv.area >= 2 {
            for d2 in 0..NUM_DIRS {
                let (dr2, dc2) = get_deltas(d2, color);
                let r2 = r1 + dr2;
                let c2 = c1 + dc2;
                if r2 < 0 || r2 >= BOARD_SIZE as i32 || c2 < 0 || c2 >= BOARD_SIZE as i32 { continue; }
                let sq2 = r2 as usize * BOARD_SIZE + c2 as usize;
                if sq2 == sq { continue; } // back to start
                let t2 = board.cells[sq2];
                if t2 != EMPTY_CELL && cell_color(t2) == color { continue; }

                if t1 != EMPTY_CELL && cell_color(t1) != color {
                    // Captured at midpoint
                    let mut m = Move::simple(sq as u16, sq2 as u16);
                    m.mid_sq = sq1 as u16;
                    m.mid_piece = cell_piece(t1);
                    m.mid_color = cell_color(t1);
                    if t2 != EMPTY_CELL {
                        m.captured_piece = cell_piece(t2);
                        m.captured_color = cell_color(t2);
                    }
                    moves.push(m);
                } else {
                    add_move(moves, sq as u16, sq2 as u16, pt, color, t2);
                }
            }
        }
    }
}

fn gen_range_capture(board: &Board, sq: usize, pt: u16, color: u8, mv: &Movement,
                     rt: &RayTable, moves: &mut Vec<Move>) {
    let piece_rank = pieces::rank(pt);

    for &dir in &mv.range_capture {
        let ray = rt.ray_for_color(sq, dir as usize, color);
        let mut captured_list: Vec<(u16, u16, u8)> = Vec::new();

        for &rsq in ray {
            let target = board.cells[rsq as usize];
            if target == EMPTY_CELL {
                let mut m = Move::simple(sq as u16, rsq);
                if !captured_list.is_empty() {
                    m.range_caps = Some(captured_list.clone());
                }
                moves.push(m);
            } else {
                let t_pt = cell_piece(target);
                let t_rank = pieces::rank(t_pt);
                if t_rank > piece_rank {
                    // Lower rank: capture and continue
                    captured_list.push((rsq, t_pt, cell_color(target)));
                    // Apply zone-based promotion logic for range capture
                    let from_in = in_promo_zone(sq, color);
                    let to_in = in_promo_zone(rsq as usize, color);
                    let may_promo = can_promote(pt) && (
                        (!from_in && to_in) || (from_in && to_in)
                    );
                    let must_promo = may_promo && is_farthest_rank(rsq as usize, color)
                        && pieces::must_promote_at_far_rank(pt);
                    if must_promo {
                        let mut m = Move::simple(sq as u16, rsq);
                        m.captured_piece = t_pt;
                        m.captured_color = cell_color(target);
                        m.range_caps = Some(captured_list.clone());
                        m.promotion = true;
                        moves.push(m);
                    } else if may_promo {
                        // Both options
                        let mut m1 = Move::simple(sq as u16, rsq);
                        m1.captured_piece = t_pt;
                        m1.captured_color = cell_color(target);
                        m1.range_caps = Some(captured_list.clone());
                        moves.push(m1);
                        let mut m2 = Move::simple(sq as u16, rsq);
                        m2.captured_piece = t_pt;
                        m2.captured_color = cell_color(target);
                        m2.range_caps = Some(captured_list.clone());
                        m2.promotion = true;
                        moves.push(m2);
                    } else {
                        let mut m = Move::simple(sq as u16, rsq);
                        m.captured_piece = t_pt;
                        m.captured_color = cell_color(target);
                        m.range_caps = Some(captured_list.clone());
                        moves.push(m);
                    }
                } else {
                    break; // blocked by equal or higher rank
                }
            }
        }
    }
}

fn gen_igui(board: &Board, sq: usize, pt: u16, color: u8, moves: &mut Vec<Move>) {
    for d in 0..NUM_DIRS {
        if let Some(nsq) = step_sq(sq, d, color) {
            let target = board.cells[nsq];
            if target != EMPTY_CELL && cell_color(target) != color {
                // Igui is a capture in place — zone promotion applies if piece is in zone
                let in_zone = in_promo_zone(sq, color);
                let may_promo = can_promote(pt) && in_zone;
                if may_promo {
                    // Both options (never forced since piece doesn't move to farthest rank)
                    let mut m1 = Move::simple(sq as u16, sq as u16);
                    m1.captured_piece = cell_piece(target);
                    m1.captured_color = cell_color(target);
                    m1.is_igui = true;
                    moves.push(m1);
                    let mut m2 = Move::simple(sq as u16, sq as u16);
                    m2.captured_piece = cell_piece(target);
                    m2.captured_color = cell_color(target);
                    m2.is_igui = true;
                    m2.promotion = true;
                    moves.push(m2);
                } else {
                    let mut m = Move::simple(sq as u16, sq as u16);
                    m.captured_piece = cell_piece(target);
                    m.captured_color = cell_color(target);
                    m.is_igui = true;
                    moves.push(m);
                }
            }
        }
    }
}
